### PR TITLE
OSASINFRA-3311: openstack: CAPO provider deployment

### PR DESCRIPTION
**Pull request by EmilienM**
**What this PR does / why we need it**:

- **openstack: add CAPO deployment and OpenStackCluster object**
- **Run `make update`**

This is the third PR to support OpenStack as dev-preview in 4.17.
This PR focuses on the CAPI provider. There are 2 commits:
* One that reviewers should review `1004 insertions(+), 146 deletions(-)`
* One that was generated by `make update`

### CODE REVIEW

**Comment by EmilienM** [comment]
/test unit
sounds like a flake in a unrelated test.

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 1]
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
...
+
+	// CloudName specifies the name of the entry in the clouds.yaml file to use.
+	// +kubebuilder:validation:Required
+	CloudName string `json:"cloudName"`
+}
+
+type SubnetSpec struct {
+	// DNSNameservers holds a list of DNS server addresses that will be provided when creating
+	// the subnet. These addresses need to have the same IP version as CIDR.
+	DNSNameservers []string `json:"dnsNameservers,omitempty"`
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/hostedcluster_types.go:1925]: Can we specifically say whether or not this is required or optional please?

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 2]
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
...
+
+	// AllocationPools is an array of AllocationPool objects that will be applied to OpenStack Subnet being created.
+	// If set, OpenStack will only allocate these IPs for Machines. It will still be possible to create ports from
+	// outside of these ranges manually.
+	AllocationPools []AllocationPool `json:"allocationPools,omitempty"`
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/hostedcluster_types.go:1930]: Can we specifically say whether or not this is required or optional please?

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 3]
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
...
+}
+
+type AllocationPool struct {
+	// Start represents the start of the AllocationPool, that is the lowest IP of the pool.
+	// +required
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/hostedcluster_types.go:1935]: Does this need `// +kubebuilder:validation:Required`?

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 4]
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
...
+	Start string `json:"start"`
+
+	// End represents the end of the AlloctionPool, that is the highest IP of the pool.
+	// +required
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/hostedcluster_types.go:1939]: Does this need `// +kubebuilder:validation:Required`?

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 5]
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
...
+// +kubebuilder:validation:MaxProperties:=1
+// +kubebuilder:validation:MinProperties:=1
+type RouterParam struct {
+	// ID is the ID of the router to use. If ID is provided, the other filters cannot be provided. Must be in UUID format.
+	// +kubebuilder:validation:Format:=uuid
+	// +optional
+	ID *string `json:"id,omitempty"`
+
+	// Filter specifies a filter to select an OpenStack router. If provided, cannot be empty.
+	Filter *RouterFilter `json:"filter,omitempty"`
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/hostedcluster_types.go:1953]: Can we specifically say whether or not this is required or optional please?

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 6]
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
...
+}
+
+// RouterFilter specifies a query to select an OpenStack router. At least one property must be set.
+// +kubebuilder:validation:MinProperties:=1
+type RouterFilter struct {
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/hostedcluster_types.go:1958]: For each property, can we specifically say whether or not this is required or optional please?

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 7]
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
...
+	ID *string `json:"id,omitempty"`
+
+	// Filter specifies a filter to select an OpenStack network. If provided, cannot be empty.
+	// +optional
+	Filter *NetworkFilter `json:"filter,omitempty"`
+}
+
+// NetworkFilter specifies a query to select an OpenStack network. At least one property must be set.
+// +kubebuilder:validation:MinProperties:=1
+type NetworkFilter struct {
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/hostedcluster_types.go:1982]: For each property, can we specifically say whether or not this is required or optional please?

File: api/hypershift/v1beta1/hostedcluster_types.go [hunk 8]
```diff
@@ -1832,8 +1836,217 @@ type AzurePlatformSpec struct {
...
+	ID *string `json:"id,omitempty"`
+
+	// Filter specifies a filter to select the subnet. It must match exactly one subnet.
+	// +optional
+	Filter *SubnetFilter `json:"filter,omitempty"`
+}
+
+// SubnetFilter specifies a filter to select a subnet. At least one parameter must be specified.
+// +kubebuilder:validation:MinProperties:=1
+type SubnetFilter struct {
```
Comment (bryan-cox) [inline api/hypershift/v1beta1/hostedcluster_types.go:2039]: For each property, can we specifically say whether or not this is required or optional please?

**Review by bryan-cox — COMMENTED** [review]
Mostly API comments. We will be going through an effort to mark all API fields as required or optional explicitly; trying to save reworking through these in the future 😄

**Comment by EmilienM** [comment]
> Mostly API comments. We will be going through an effort to mark all API fields as required or optional explicitly; trying to save reworking through these in the future 😄

Thanks, great feedback. We'll keep that in mind for our future API changes.
I think I addressed the comments.

**Pull request merged**